### ISSUE TITLE
System tests

### ISSUE THREAD

**Issue by nkuba** [keep-network/keep-ecdsa#382]
# Bonded ECDSA Keep Test Cases

## Random Beacon Integration

* [ ] Call openKeepFeeEstimate and openKeep when Random Beacon clients are functional.
* [ ] Call openKeepFeeEstimate and openKeep when Random Beacon clients are not working.
* [ ] Batch calls to Random Beacon: `openKeepFeeEstimate` and `openKeep` called `n` times quickly one after another
* [ ] Reseed BondedECDSAKeepFactory when there are no funds in the reseed pool
* [ ] Reseed BondedECDSAKeepFactory when there is enough funds in the pool to reseed
with no additional payment.
* [ ] Reseed BondedECDSAKeepFactory with payment lower than estimated fee.

### Positive scenarios

* [ ] Request with minimum group size (1) and threshold = group size
* [ ] Request with group size = 3 and threshold = group size
* [ ] Request with maximum group size (16) and threshold = group size
* [ ] Request with threshold = 1
* [ ] Request with threshold = group size - 1
* [ ] Request with threshold = group size
* [ ] Request `n` keeps at once:

  * `n` = 10
  * `n` = 100
  * `n` = 1000

* [ ] Request keeps for two different Sanctioned Applications supported by at least
one client at once
* [ ] Signers that are not part of the keep ignore the request

Expected results:
- members keys submitted
- public key set for keep
- data stored in storage

### Negative scenarios

* [ ] Request with group size = 0
* [ ] Request with threshold = 0
* [ ] Request with threshold = group size + 1
* [ ] Call from not supported Sanctioned Application
* [ ] Request for group size greater than the number of registered clients

### Client Unavailability

* [ ] Request when one of the clients selected to the group is down and
threshold = group size
* [ ] Request when one of the clients selected to the group is down and
threshold = group size - 1
* [ ] Temporarily break a client while generating a group key when threshold = group size
* [ ] Permanently break a client while generating a group key when threshold = group size

### Positive scenarios

* [ ] Request signature from keep with group size = 1
* [ ] Request signature from keep with group size = 3
* [ ] Request signature from keep with maximum group size (16)
* [ ] Request signatures from `n` keeps at once:
  * `n` = 10
  * `n` = 100
  * `n` = 1000

* [ ] Request signatures for two different Sanctioned Applications from keeps 
supported by the same client
* [ ] Signers that are not part of the keep ignore the request

Expected results:
- clients calculate a signature
- only one client submits a signature
- event is emitted

### Negative scenarios

* [ ] Request signature before the public key is generated
* [ ] Request signature from a closed keep
* [ ] Multiple concurrent executions for the same keep

### Client Unavailability

* [ ] Request when one of the signers is down and threshold = group size
* [ ] Request when one of the clients selected to the group is down and
threshold = group size - 1
* [ ] Temporarily break a client while generating a group key when threshold = group size
* [ ] Permanently break a client while generating a group key when threshold = group size

## Close Keep

* [ ] Call closeKeep by the keep owner

## Seize Signer Bonds

* [ ] Call seizeSignerBonds by the keep owner

## Submit Signature Fraud

To test these scenarios easier we should expect a network of 3 operator nodes
and keeps with a group size of 3.

Below tests should be executed for two situations:
1. operator is supporting just one keep
1. operator is supporting multiple keeps

* [ ] Submit valid signature fraud when one of keeps members operators has just
a minimum stake
* [ ] Submit valid signature fraud for operators that have multiplies of the minimum stake

# Client Test Cases

## Client Configuration

Start client with:
* [ ] valid configuration
* [ ] missing config file
* [ ] config file in default location
* [ ] config file in custom location
* [ ] invalid formatted config file
* [ ] invalid ETH URL
* [ ] invalid ETH URLRPC
* [ ] missing ethereum key file
* [ ] invalid ethereum key file
* [ ] no ethereum password environment variable provided
* [ ] incorrect ethereum password environment variable provided
* [ ] after some keeps are open restart client with incorrect ethereum password
* [ ] missing BondedECDSAKeepFactory contract address
* [ ] BondedECDSAKeepFactory contract address of other contract which is not the factory
* [ ] BondedECDSAKeepFactory contract address of old factory
* [ ] invalid sanctioned application ethereum address
* [ ] not registered sanctioned application address
* [ ] one sanctioned application address
* [ ] missing storage path
* [ ] storage path where last directory doesn't exist
* [ ] storage path where `current` and `archive` directories don't exist
* [ ] storage path already containing some keep data
* [ ] empty LibP2P peers list
* [ ] address where LibP2P peer is not working
* [ ] one LibP2P peer address
* [ ] LibP2P peer that is running different client version
* [ ] multiple peer addresses
* [ ] multiple peer addresses where one is not working
* [ ] LibP2P port that is already in use
* [ ] LibP2P port behind firewall for in traffic
* [ ] LibP2P port behind firewall for in/out traffic
* [ ] LibP2P port behind firewall for out traffic
* [ ] LibP2P announced address not provided
* [ ] LibP2P announced incorrect address provided
* [ ] LibP2P announced address dns4 provided
* [ ] LibP2P announced address ip4 provided
* [ ] LibP2P announced address dns4 + ip4 provided

## Multiple Clients

* [ ] Run min number of clients (1)
* [ ] Run `n` clients
  * `n` = 10
  * `n` = 100
  * `n` = 1000

* [ ] New nodes joining continuously and are getting picked for new groups randomly
* [ ] Run two groups of clients supporting one address of BondedECDSAKeepFactory
and the second group supporting another address of BondedECDSAKeepFactory.
Submit some open keep and sign requests to both factory contracts.

# Deployments Test Cases

## Initial Migration

* [ ] Migrate all contracts and verify gas and ether used.

## Upgrades

* [ ] Upgrade BondedECDSAKeepVendor
* [ ] Upgrade BondedECDSAKeepFactory

# Operator Test Cases

These tests are related to a single instance of the client. They are not required 
to be executed for different network sizes.

## Stake

* [ ] Start client when operator account has just minimum stake
* [ ] Start client when operator account has multiples of minimum stake

### Insufficient stake

* [ ] Start client when operator account has ZERO stake
* [ ] Start client when operator account has below minimum stake
* [ ] Drop stake below minimum while client is running
* [ ] Drop stake below minimum and restart the client

## Ether

* [ ] Start client when operator account has some ETH
* [ ] Drop ETH for operator account while client is running
* [ ] Drop ETH for operator account and restart the client
* [ ] Start client when operator account has ZERO ETH
* [ ] Monitor ETH usage for opening new keep and generating a signature

## Bond

* [ ] Start client when operator account has ZERO unbonded value
* [ ] Start client when operator account has 1 ETH of unbonded value
* [ ] Request new keep when operator's unbonded value is ZERO
* [ ] Request new keep when operator's unbonded value is positive but lower than requested
bond / group size
* [ ] Request new keep when operator's unbonded value is equal to requested bond / group size
* [ ] Request new keep when operator's unbonded value is greater than requested bond / group size
* [ ] Request a signature for existing keep when operators unbonded value is ZERO

## Operator Contract Authorization

* [ ] Start client without BondedECDSAKeepFactory authorization in TokenStaking contract
* [ ] Authorize BondedECDSAKeepFactory in TokenStaking contract for operator

## Sortition Pool Authorization

* [ ] Start client when sortition pool is not yet created
* [ ] Start client without sortition pool authorization in KeepBonding contract
* [ ] Authorize sortition pool in KeepBonding contract for operator

## Stake Initialization Period

Execute these tests in a setup where group size is the same as number of running
nodes.

* [ ] Start client just after delegating a stake to the operator, so stake initialization
period has not passed yet.
* [ ] Request new keep after stake initialization period (10 blocks) has passed

**Comment by nkuba** [comment]
Deployment cost: 13,230,733 gas

<details>
# Migrations dry-run (simulation)

> Network name: 'ropsten-fork'
> Network id: 3
> Block gas limit: 0x7a1200

# 1_initial_migration.js

Deploying 'Migrations'

---

> block number: 7684574
> block timestamp: 1586446766
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.947066740976362429
> gas used: 210237
> gas price: 2 gwei
> value sent: 0 ETH
> total cost: 0.000420474 ETH

---

> Total cost: 0.000420474 ETH

# 2_deploy_contracts.js

Browserslist: caniuse-lite is outdated. Please run next command `npm update`

Deploying 'Branch'

---

> block number: 7684576
> block timestamp: 1586446777
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.946868586976362429
> gas used: 71714
> gas price: 2 gwei
> value sent: 0 ETH
> total cost: 0.000143428 ETH

Deploying 'Position'

---

> block number: 7684577
> block timestamp: 1586446780
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.946725158976362429
> gas used: 71714
> gas price: 2 gwei
> value sent: 0 ETH
> total cost: 0.000143428 ETH

Deploying 'StackLib'

---

> block number: 7684578
> block timestamp: 1586446783
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.946523076976362429
> gas used: 101041
> gas price: 2 gwei
> value sent: 0 ETH
> total cost: 0.000202082 ETH

Deploying 'Leaf'

---

> block number: 7684579
> block timestamp: 1586446785
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.946379648976362429
> gas used: 71714
> gas price: 2 gwei
> value sent: 0 ETH
> total cost: 0.000143428 ETH

Linking

---

- Contract: BondedSortitionPoolFactory <--> Library: StackLib

Deploying 'BondedSortitionPoolFactory'

---

> block number: 7684580
> block timestamp: 1586446788
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.941257382976362429
> gas used: 2561133
> gas price: 2 gwei
> value sent: 0 ETH
> total cost: 0.005122266 ETH

Deploying 'KeepBonding'

---

> block number: 7684581
> block timestamp: 1586446795
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.937748596976362429
> gas used: 1754393
> gas price: 2 gwei
> value sent: 0 ETH
> total cost: 0.003508786 ETH

Deploying 'BondedECDSAKeep'

---

> block number: 7684582
> block timestamp: 1586446800
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.930252988976362429
> gas used: 3747804
> gas price: 2 gwei
> value sent: 0 ETH
> total cost: 0.007495608 ETH

Deploying 'BondedECDSAKeepFactory'

---

> block number: 7684583
> block timestamp: 1586446816
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.924732158976362429
> gas used: 2760415
> gas price: 2 gwei
> value sent: 0 ETH
> total cost: 0.00552083 ETH

Deploying 'BondedECDSAKeepVendorImplV1'

---

> block number: 7684584
> block timestamp: 1586446829
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.922939790976362429
> gas used: 896184
> gas price: 2 gwei
> value sent: 0 ETH
> total cost: 0.001792368 ETH

Deploying 'BondedECDSAKeepVendor'

---

> block number: 7684585
> block timestamp: 1586446848
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.921030548976362429
> gas used: 954621
> gas price: 2 gwei
> value sent: 0 ETH
> total cost: 0.001909242 ETH

---

> Total cost: 0.025981466 ETH

# 3_initialize.js

current factory address: [0x128Fe4c78Bd23D33d30EBc5a42bC49A1D3659786]
approved operator contract [0x128Fe4c78Bd23D33d30EBc5a42bC49A1D3659786] in registry
set operator [0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D] as [0x9b873013CC55Ab1dF094B19d46EB1DaE608b7D81] contract upgrader

---

> Total cost: 0 ETH

# Summary

> Total deployments: 11
> Final cost: 0.02640194 ETH

# Starting migrations...

> Network name: 'ropsten'
> Network id: 3
> Block gas limit: 0x7a121d

# 1_initial_migration.js

Deploying 'Migrations'

---

> transaction hash: 0x976e4ba852510de42cec32a39e78d02538c091f0df6c31c6c0395f8037fa6da8
> Blocks: 2 Seconds: 44
> contract address: 0xE78e9f947328E28cD3c86e229b8Fc176Ac5e7A7f
> block number: 7684583
> block timestamp: 1586446885
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.942982474976362429
> gas used: 225237
> gas price: 20 gwei
> value sent: 0 ETH
> total cost: 0.00450474 ETH

> Saving migration to chain.
> Saving artifacts

---

> Total cost: 0.00450474 ETH

# 2_deploy_contracts.js

Deploying 'Branch'

---

> transaction hash: 0xc913fa483bc7743a8d5d5256d82e73f700e465175add9e57739d1ea96d8db9da
> Blocks: 2 Seconds: 12
> contract address: 0xd11699f7CF0a88BEa8696DDD94F827E6007256c0
> block number: 7684587
> block timestamp: 1586446974
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.940700934976362429
> gas used: 71714
> gas price: 20 gwei
> value sent: 0 ETH
> total cost: 0.00143428 ETH

Deploying 'Position'

---

> transaction hash: 0xad56c4419284fccc1b95735c11f71c75a8359b6e701e8676629ed6aa03d17a8a
> Blocks: 1 Seconds: 16
> contract address: 0x3c3225d8ef761e66519F353254a60e380D52D303
> block number: 7684590
> block timestamp: 1586446986
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.947668568170806873
> gas used: 71714
> gas price: 20 gwei
> value sent: 0 ETH
> total cost: 0.00143428 ETH

Deploying 'StackLib'

---

> transaction hash: 0xbfda945edd7271764289aa2c0cdbbda346a9a7efeda64ab82a9d713040b011d1
> Blocks: 1 Seconds: 12
> contract address: 0x685a1552921DefEe39A23d9Fef9167D490442AeF
> block number: 7684592
> block timestamp: 1586447011
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.945647748170806873
> gas used: 101041
> gas price: 20 gwei
> value sent: 0 ETH
> total cost: 0.00202082 ETH

Deploying 'Leaf'

---

> transaction hash: 0x3ecbb5399b02865e1bf2856df5531de51827844a4c41732559c12c45b34bd67b
> Blocks: 1 Seconds: 60
> contract address: 0xe80CF483070348eD6DAc2743C04E97dc0Bc68FAa
> block number: 7684595
> block timestamp: 1586447065
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.944213468170806873
> gas used: 71714
> gas price: 20 gwei
> value sent: 0 ETH
> total cost: 0.00143428 ETH

Linking

---

- Contract: BondedSortitionPoolFactory <--> Library: StackLib (at address: 0x685a1552921DefEe39A23d9Fef9167D490442AeF)

Deploying 'BondedSortitionPoolFactory'

---

> transaction hash: 0x5d39170d0eb3a70d39ae33caf49e88a46ef2fa41d758d3eb22af1cb5b0abd666
> Blocks: 1 Seconds: 52
> contract address: 0x1E9cd46DE58b2e29009b40B6f490a0E6edAd590B
> block number: 7684597
> block timestamp: 1586447124
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.892990808170806873
> gas used: 2561133
> gas price: 20 gwei
> value sent: 0 ETH
> total cost: 0.05122266 ETH

Deploying 'KeepBonding'

---

> transaction hash: 0xf92e2d9df5a63e0eb77473879a2d21538c1d6d15c3e3ceb0567f97e5e4c0954a
> Blocks: 1 Seconds: 24
> contract address: 0x33AD3A082871943161B89cADEf69a6729542F77C
> block number: 7684599
> block timestamp: 1586447156
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.857302948170806873
> gas used: 1784393
> gas price: 20 gwei
> value sent: 0 ETH
> total cost: 0.03568786 ETH

Deploying 'BondedECDSAKeep'

---

> transaction hash: 0x60c096648c321b780a233e8eb45820800c9a41505ca3c8bf4a9e6799072b8a86
> Blocks: 2 Seconds: 68
> contract address: 0xeA40eBf57f615684129B0a7eefe56FF27393635e
> block number: 7684603
> block timestamp: 1586447210
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.782346868170806873
> gas used: 3747804
> gas price: 20 gwei
> value sent: 0 ETH
> total cost: 0.07495608 ETH

Deploying 'BondedECDSAKeepFactory'

---

> transaction hash: 0x163550cc6a59a99cd1967c33e5c369a87d5d63cfa3f6dbf7b09845a4f486c228
> Blocks: 0 Seconds: 4
> contract address: 0x128Fe4c78Bd23D33d30EBc5a42bC49A1D3659786
> block number: 7684605
> block timestamp: 1586447261
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.725338568170806873
> gas used: 2850415
> gas price: 20 gwei
> value sent: 0 ETH
> total cost: 0.0570083 ETH

Deploying 'BondedECDSAKeepVendorImplV1'

---

> transaction hash: 0xc4e2115440147443ef2649b73a1ee5d85d427a47e1590e335dd2abde8d5c93b8
> Blocks: 1 Seconds: 16
> contract address: 0x0e6626189d11E1eED23De18e525Bf8B414B8Bd4e
> block number: 7684607
> block timestamp: 1586447280
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.707114888170806873
> gas used: 911184
> gas price: 20 gwei
> value sent: 0 ETH
> total cost: 0.01822368 ETH

Deploying 'BondedECDSAKeepVendor'

---

> transaction hash: 0x1099122bfd0ff5bed9e10ec687e205ccf45b4b65d9db3536f3d8e43aa1aa7608
> Blocks: 1 Seconds: 12
> contract address: 0x9b873013CC55Ab1dF094B19d46EB1DaE608b7D81
> block number: 7684610
> block timestamp: 1586447302
> account: 0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D
> balance: 261.685922468170806873
> gas used: 1059621
> gas price: 20 gwei
> value sent: 0 ETH
> total cost: 0.02119242 ETH

> Saving migration to chain.
> Saving artifacts

---

> Total cost: 0.26461466 ETH

# 3_initialize.js

current factory address: [0x128Fe4c78Bd23D33d30EBc5a42bC49A1D3659786]
approved operator contract [0x128Fe4c78Bd23D33d30EBc5a42bC49A1D3659786] in registry
set operator [0x923C5Dbf353e99394A21Aa7B67F3327Ca111C67D] as [0x9b873013CC55Ab1dF094B19d46EB1DaE608b7D81] contract upgrader

> Saving migration to chain.

---

> Total cost: 0 ETH

# Summary

> Total deployments: 11
> Final cost: 0.2691194 ETH

</details>
